### PR TITLE
[NetworkInterface] Added module

### PR DESCRIPTION
**Pull request by MrMCake**
# Description

- Added deciated Network Interface module

| Pipeline |
| - |
|[Build Status](https://dev.azure.com/servicescode/infra-as-code-source/_build/latest?definitionId=2399&branchName=users%2Falsehr%2F1396_nic) |

# Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update (Wiki)

### CODE REVIEW

File: arm/.global/global.module.tests.ps1 [hunk 1]
```diff
@@ -49,8 +49,8 @@ Describe 'File/folder tests' -Tag Modules {
 
                 $workflowsFolderName = Join-Path $RepoRoot '.github' 'workflows'
                 $workflowFileName = '{0}.yml' -f $moduleFolderName.Replace('\', '/').Replace('/', '.').Replace('Microsoft', 'ms').ToLower()
-
-                Test-Path (Join-Path $workflowsFolderName $workflowFileName) | Should -Be $true
+                $workflowPath = Join-Path $workflowsFolderName $workflowFileName
```
Comment (eriqua) [inline arm/.global/global.module.tests.ps1:52]: Is this related to this PR?

File: arm/Microsoft.Network/networkInterfaces/.parameters/parameters.json [hunk 2]
```diff
@@ -0,0 +1,48 @@
+{
```
Comment (eriqua) [inline arm/Microsoft.Network/networkInterfaces/.parameters/parameters.json:1]: Please add the min.parameters.json

File: arm/.global/global.module.tests.ps1 (hunk 1 above)
Comment (MrMCake) [reply arm/.global/global.module.tests.ps1:52]: Yes. The teast wasn't as meaningful as it should have been

File: arm/Microsoft.Network/networkInterfaces/.parameters/parameters.json (hunk 2 above)
Comment (MrMCake) [reply arm/Microsoft.Network/networkInterfaces/.parameters/parameters.json:1]: done

File: .github/workflows/ms.network.networkinterfaces.yml [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+      - 'arm/Microsoft.Network/networkInterfaces/**'
+      - 'arm/.global/global.module.tests.ps1'
+      - '!*/**/readme.md'
+      - 'utilities/pipelines/**'
+      - '!utilities/pipelines/dependencies/**'
+
+env:
+  variablesPath: 'global.variables.yml'
+  modulePath: 'arm/Microsoft.Network/networkInterfaces'
+  workflowPath: '.github/workflows/ms.network.networkInterfaces.yml'
```
Comment (eriqua) [inline .github/workflows/ms.network.networkinterfaces.yml:31]: Currently all workflow file names are lowercase so I'd suggest to rename the file and then this path for consistency (Remember the change of case only won't be taken as an update. You'd need to update the name in git twice)

File: .github/workflows/ms.network.networkinterfaces.yml [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        type: boolean
+        description: 'Publish prerelease module'
+        required: false
+        default: false
+  push:
+    branches:
+      - main
+    paths:
+      - '.github/actions/templates/**'
+      - '.github/workflows/ms.network.networkInterfaces.yml'
```
Comment (eriqua) [inline .github/workflows/ms.network.networkinterfaces.yml:21]:
```suggestion
      - '.github/workflows/ms.network.networkinterfaces.yml'
```

File: .azuredevops/modulePipelines/ms.network.networkinterfaces.yml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+pr: none
+
+trigger:
+  batch: true
+  branches:
+    include:
+      - main
+  paths:
+    include:
+      - '/.azuredevops/modulePipelines/ms.network.networkInterfaces.yml'
```
Comment (eriqua) [inline .azuredevops/modulePipelines/ms.network.networkinterfaces.yml:22]:
```suggestion
      - '/.azuredevops/modulePipelines/ms.network.networkinterfaces.yml'
```

File: arm/Microsoft.Network/networkInterfaces/deploy.bicep [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        gatewayLoadBalancer: contains(ipConfiguration, 'gatewayLoadBalancer') ? ipConfiguration.gatewayLoadBalancer : null
+        loadBalancerInboundNatRules: contains(ipConfiguration, 'loadBalancerInboundNatRules') ? ipConfiguration.loadBalancerInboundNatRules : null
+        privateIPAddressVersion: contains(ipConfiguration, 'privateIPAddressVersion') ? ipConfiguration.privateIPAddressVersion : null
+        virtualNetworkTaps: contains(ipConfiguration, 'virtualNetworkTaps') ? ipConfiguration.virtualNetworkTaps : null
+      }
+    }]
+  }
+}
+
+resource publicIpAddress_diagnosticSettings 'Microsoft.Insights/diagnosticSettings@2021-05-01-preview' = if (!empty(diagnosticStorageAccountId) || !empty(diagnosticWorkspaceId) || !empty(diagnosticEventHubAuthorizationRuleId) || !empty(diagnosticEventHubName)) {
```
Comment (eriqua) [inline arm/Microsoft.Network/networkInterfaces/deploy.bicep:126]:
```suggestion
resource networkInterface_diagnosticSettings 'Microsoft.Insights/diagnosticSettings@2021-05-01-preview' = if (!empty(diagnosticStorageAccountId) || !empty(diagnosticWorkspaceId) || !empty(diagnosticEventHubAuthorizationRuleId) || !empty(diagnosticEventHubName)) {
```

File: arm/Microsoft.Network/networkInterfaces/deploy.bicep [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+        applicationGatewayBackendAddressPools: contains(ipConfiguration, 'applicationGatewayBackendAddressPools') ? ipConfiguration.applicationGatewayBackendAddressPools : null
+        gatewayLoadBalancer: contains(ipConfiguration, 'gatewayLoadBalancer') ? ipConfiguration.gatewayLoadBalancer : null
+        loadBalancerInboundNatRules: contains(ipConfiguration, 'loadBalancerInboundNatRules') ? ipConfiguration.loadBalancerInboundNatRules : null
+        privateIPAddressVersion: contains(ipConfiguration, 'privateIPAddressVersion') ? ipConfiguration.privateIPAddressVersion : null
+        virtualNetworkTaps: contains(ipConfiguration, 'virtualNetworkTaps') ? ipConfiguration.virtualNetworkTaps : null
+      }
+    }]
+  }
+}
+
```
Comment (eriqua) [inline arm/Microsoft.Network/networkInterfaces/deploy.bicep:125]: What about cross-referenced pip configurations?

Comment (MrMCake) [reply arm/Microsoft.Network/networkInterfaces/deploy.bicep:125]: Not sure I follow

Comment (eriqua) [reply arm/Microsoft.Network/networkInterfaces/deploy.bicep:125]: ref how nics are developed for vms

Comment (MrMCake) [reply arm/Microsoft.Network/networkInterfaces/deploy.bicep:125]: Will be handled by #1410

**Review by eriqua — APPROVED** [review]
🆗
